### DISCUSSION TITLE
How can I "turn back time" and go back to a previous tag/release of Eclipse?

### DISCUSSION THREAD

**Discussion by fedejeanne** [eclipse-platform/eclipse.platform.releng.aggregator#1403] [Q&A] [answered]
## <TL;DR>
How do I develop in a maintenance branch?

## The long version
Hello everyone,
I'm trying to go back to a previous release of Eclipse (`4.29`) in order to properly debug one specific application via _remote debug_ .

I tried checking out the _branch_ `R4_29_maintenance` in all my repos (those which actually have that branch) in the WS ...

![image](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/assets/2205684/47ae6039-839f-494a-99c7-9701cda1862f)

... and then I opened `eclipse-sdk-prereqs.target` and let it resolve its dependencies...

![image](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/assets/2205684/e5e7474b-f257-4923-b663-5279d03a8ac7)

...but it didn't work out of the box: I'm having trouble getting the right **Target platform** even after a restart:

(After a restart)
![image](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/assets/2205684/ff587bef-39e0-41ab-a139-2f4494c0fe69)

(Also after a restart)
![image](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/assets/2205684/c92ec55d-a5a3-42f9-86a2-63be3466c78a)

In the last screenshot you can see that some references are still pointing to `4.30`. 

What am I missing here?

**Comment by mickaelistria** [comment]
In order to debug an existing application, one more direct possibility is to set up a target platform that does reference the installation path of the application you're debugging; and just import the code of the bundles you'll inspect (using the right tag).
Or, maybe even simpler, once Target Platform for this RCP is setup, go to the Plug-ins view, find the bundle you want, and right-click > Import As > Import project from repository...

**Comment by merks** [comment, marked as answer]
The last screen shows what's being used for the installation itself, i.e., for the **p2 director** task.  It should be fine to use 4.30 for debugging with a 4.29 target platform. 

But you're also running the modular targlet tasks and those are all hard coded to use the latest of everything.

You can use Navigate -> Open Setup -> Workspace and textually copy the following and paste it onto the Workspace object and save:
```
<?xml version="1.0" encoding="UTF-8"?>
<xmi:XMI xmi:version="2.0"
    xmlns:xmi="http://www.omg.org/XMI"
    xmlns:setup="http://www.eclipse.org/oomph/setup/1.0">
  <setup:VariableTask
      name="eclipse.target.platform"
      value="2023-09"
      storageURI="scope://Workspace"/>
  <setup:VariableTask
      name="eclipse.api.baseline.target.platform"
      value="2023-06"
      storageURI="scope://Workspace"/>
</xmi:XMI>
```
This will override the values in the setups.  Then you should get the older API baseline:

![image](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/assets/208716/bcc5bd43-ad01-412d-a85a-79b49031cdd1)

and the older target platform:

![image](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/assets/208716/cb4a1a9e-a58f-4681-a9e6-40ee4bf002ee)

**Comment by fedejeanne** [comment]
Thank you both for your answers!
@mickaelistria you're right, in most cases it suffices to simply get the right code as "read-only" and debug with it, but in this cases I also wanted to be able to tweak the code here and there which is why I wanted to download and compile the targeted version.

@merks thank you, I changed the configuration and was able to compile and run the code. I've a final question though: in your snippet you set the **Modular Target** (`eclipse.target.platform`) to 2023-09 and the **Modular API Baseline Target** (`eclipse.api.baseline.target.platform`) to 2023-06. Was this intentional? 

Since I was targeting 4.29, I just set both to 2023-09 and it worked.

**Reply by merks** [reply]
It's intentional.  Currently these values (from the projects catalog) are used:

![image](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/assets/208716/4efaa06b-3488-4c44-ab37-8065dab6a7f9)

The target platform is targeting the current not-yet-released versions of any dependencies.  On the other hand, the API baseline is targeting the already-released versions because of course API changes are relative to what was most-recently released....

**Reply by laeubi** [reply]
> but in this cases I also wanted to be able to tweak the code here and there which is why I wanted to download and compile the targeted version

You can still use what was suggested by @mickaelistria I won't bother using the modular target as it really complicates things more than helping here, so you can simply add the release URL and import any project you want from the git and then switch to the Tag/branch/commit/... you want.
